### DISCUSSION TITLE
How will mappers work with BATCH messaging?

### DISCUSSION THREAD

**Discussion by tayloramurphy** [meltano/meltano#6639] [General]
SDK discussion link:

- https://github.com/meltano/sdk/issues/1117

One concern I brought up with @aaronsteers today was how stream maps will work with the `BATCH` message type. 

My desired end state is that a mapper built on the SDK works with batch workflows out of the box no problem. 

A worst case scenario is that a user is expecting something to use batch and they throw a mapper in there and it reverts back to single messages in a way that's non-transparent. 

Do we know if SDK-based stream maps will work out of the box? Do we need to make decisions about defaults around file cleanup / overwriting for the mappers?

@aaronsteers @edgarrmondragon

**Comment by edgarrmondragon** [comment]
@tayloramurphy 

> Do we know if SDK-based stream maps will work out of the box?

Not really. For dumping batches of records from an API that's already returning individual records, which the SDK can simply _convert_ into a batch file, streams maps configured for the tap (e.g. `{"token": ..., "stream_maps": {...}}`) could work as they do on normal sync.

However, for systems with builtin dumping/unloading functionality like most cloud data warehouses, the way stream maps operate on a record-by-record basis would defeat the purpose of BATCH _or_ require the stream map to do a lot of (inefficient) work that might require different access levels not used by the tap:

1. read the batch file from the filesystem (disk, S3, GCS, etc.)
2. decode the records from JSONL, CSV, etc.
3. transform each record
4. encode the records in the exact same format as the original file
5. write the new file back to the filesystem

It's similar case for Meltano mapper packages _built_ on the SDK, such as `meltano-map-transform`.

> Do we need to make decisions about defaults around file cleanup / overwriting for the mappers?

Yes, and this is an easier question to answer: I'd prefer cleanup of the originals + emitting a new batch message that points to the new files.

**Reply by aaronsteers** [reply]
> For dumping batches of records from an API that's already returning individual records, which the SDK can simply convert into a batch file, streams maps configured for the tap (e.g. {"token": ..., "stream_maps": {...}}) could work as they do on normal sync.

Agreed. 👍 

> However, for systems with builtin dumping/unloading functionality like most cloud data warehouses, the way stream maps operate on a record-by-record basis would defeat the purpose of BATCH or require the stream map to do a lot of (inefficient) work that might require different access levels not used by the tap

Agreed here also. It is _extremely likely_ that the extra overhead would defeat the purpose of a native bulk unload operation. My counterpoint argument below is that _technically_ they can still work together and also _sometimes_ the network/handshake cost of per-record operations may still outweigh the relative cost of serial read/transform/write operations. (Possible but admittedly not a sure thing by any stretch.)

**Comment by aaronsteers** [comment]
> My desired end state is that a mapper built on the SDK works with batch workflows out of the box no problem.

I think, to maybe reconcile @edgarrmondragon and my differing responses here, a lot is wrapped up in our definition of "no problem".

If we just mean that the process can finish without failure, and while still performing the requested transformation(s), then yes, I think we're basically okay here. Important to note: **yes** we do likely lose the performance benefits of batching. 

The extent of the lost performance will depend on the relative efficiency of native BULK UNLOAD vs serial record-by-record processing from the source. If the BULK UNLOAD is only marginally faster, then the net is likely to be much slower. If the source does not support BULK UNLOAD natively, there isn't much penalty. If the BULK UNLOAD is 10x or 100x the efficiency of per-record operations (such as with Redshift), then the resulting performance _might_ still be significantly faster. (Benchmarking would have to prove this out.)

With zero optimizations we lose some of the benefits of batching but in theory this still works...

1. The tap developer decides on one of the two strategies:
   - Allow the mapping to process the records before serializing to BATCH.
     - This is the default/automatic behavior and the only one available if the source doesn't natively have something like a BULK UNLOAD capability.
     - This would also be the forced behavior (_instead of a native BULK UNLOAD_) until we do additional work to accomodate the below.
   - Allow the source system to BULK UNLOAD the data, then deserialize and process the rows using the provided mapping logic, and finally serialize a transformed data to a new set of files.
2. The target gets the BATCH files as expected.

I think it's important to note that while this does _feel_ very inefficient, and sometimes it will be much worse performance, there can be cases where the savings of being able to parallelize the work will still outweigh the inefficiencies of reading and redundantly re-writing the files. If comparing against a process which performs a full round trip to the remote API for _each record_, or for every 10 records (for instance), this could still end up being a lot faster - since serial (non-random) reads and writes to/from disk can happen very quickly - relative to per-record communications across the network.

Also, to point out a couple more things:

- If the source doesn't have any native BULK UNLOAD capability anyway, there really isn't much conflict between BATCH and stream maps - as long as they both are performed natively in the tap. 
- If an independent mapper is put between a tap and target using BATCH records, with the goal of transforming records inline, then all of the disadvantages and inefficiencies noted above will be present here as well.
  - Hopefully there aren't many cases of this, since the both BATCH and Stream Maps are built into the SDK. Meaning, taps that have BATCH capability should also have Stream Maps capability.

**Reply by tayloramurphy** [reply]
@aaronsteers thanks for the write-up. This is all helpful and it makes me think that we'll just need to have some docs and FAQs on how stream maps work with batch-based replication. Generally we'll want to push folks to use SDK-based connectors so the stream map is more performant, but I wouldn't want to block users from using independent mappers either.
